hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -1225,130 +1301,292 @@ private boolean innerRename(Path source, Path dest)
-            createObjectAttributes(status.getPath(),
-                status.getETag(), status.getVersionId());
-        S3AReadOpContext readContext = createReadContext(status, inputPolicy,
-            changeDetectionPolicy, readAhead);
-        CopyResult copyResult = copyFile(key, newDstKey, length,
-            objectAttributes, readContext);
-
-        if (hasMetadataStore()) {
-          // with a metadata store, the object entries need to be updated,
-          // including, potentially, the ancestors
-          Path childSrc = keyToQualifiedPath(key);
-          Path childDst = keyToQualifiedPath(newDstKey);
-          if (objectRepresentsDirectory(key, length)) {
-            S3Guard.addMoveDir(metadataStore, srcPaths, dstMetas, childSrc,
-                childDst, username);
-          } else {
-            S3Guard.addMoveFile(metadataStore, srcPaths, dstMetas, childSrc,
-                childDst, length, getDefaultBlockSize(childDst), username,
-                copyResult.getETag(), copyResult.getVersionId());
+
+        Path parentPath = keyToQualifiedPath(srcKey);
+        final RemoteIterator<S3ALocatedFileStatus> iterator =
+            listFilesAndEmptyDirectories(parentPath, true);
+        while (iterator.hasNext()) {
+          S3ALocatedFileStatus status = iterator.next();
+          String k = pathToKey(status.getPath());
+          String key = (status.isDirectory() && !k.endsWith("/"))
+              ? k + "/"
+              : k;
+          String newDstKey =
+              dstKey + key.substring(srcKey.length());
+          Path childSourcePath = keyToQualifiedPath(key);
+
+          queueToDelete.apply(childSourcePath, key);
+
+          Path childDestPath = keyToQualifiedPath(newDstKey);
+          S3ObjectAttributes sourceAttributes =
+              createObjectAttributes(
+                  status.getPath(),
+                  status.getETag(),
+                  status.getVersionId(),
+                  status.getLen());
+          S3AReadOpContext readContext = createReadContext(status, inputPolicy,
+              changeDetectionPolicy, readAhead);
+          // queue the copy operation for execution in the thread pool
+          CompletableFuture<Path> copy = submit(boundedThreadPool, () ->
+              copySourceAndUpdateTracker(
+                  renameTracker,
+                  childSourcePath,
+                  key,
+                  sourceAttributes,
+                  readContext,
+                  childDestPath,
+                  newDstKey,
+                  true));
+          bytesCopied.addAndGet(srcStatus.getLen());
+          activeCopies.add(copy);
+          if (activeCopies.size() == renameParallelLimit) {
+            // the limit of active copies has been reached;
+            // wait for completion or errors to surface.
+            LOG.debug("Waiting for active copies to complete");
+            completeActiveCopies.apply("batch threshold reached");
```

Comment:
yes, a refactoring would be much needed here. this part is getting huge. it's harder to review and later to debug.

Reply:
Gabor: This is my attempt to clean it up. The rename logic has all been pulled out of S3AFS into the trackers and I'm defining a couple of lambda expressions in the method to avoid copy and paste duplication and effort.

The only way to do anything else would be to pull the entire code "somewhere", but because of all the calls into S3AFS, that will only make sense as part of a relayering of ops: we'd have this to work at the "object store + S3Guard layer", rather than through the FS APIs. That's a major change and not going to happen in a patch which is big enough as it is.